lib/marin/src/marin/rl/health_monitor.py

```diff
@@ -0,0 +1,591 @@
+
+    HEALTHY = "healthy"
+    WARNING = "warning"
+    CRITICAL = "critical"
+
+
+@dataclass
+class HealthWarning:
+    """A warning about environment health."""
+
+    type: WarningType
+    env_id: str
+    message: str
+    severity: str  # "low", "medium", "high", "critical"
+    timestamp: float = field(default_factory=time.time)
+    metrics: dict[str, Any] = field(default_factory=dict)
+
+
+@dataclass
+class EnvironmentHealthMetrics:
+    """Health metrics for a single environment."""
+
+    env_id: str
+    current_state: str  # "locked", "active", "graduated"
+
+    # Performance tracking
+    current_train_performance: float = 0.0
+    current_eval_performance: float = 0.0
+    peak_train_performance: float = 0.0
+    peak_eval_performance: float = 0.0
+    performance_trend: str = "stable"  # "improving", "stable", "regressing"
+
+    # State duration
+    time_in_current_state: int = 0
+    total_training_steps: int = 0
+    steps_since_last_progress: int = 0
+
+    # Graduation tracking
+    graduation_performance: float | None = None
+    graduation_step: int | None = None
+
+    # Health indicators
+    health_status: HealthStatus = HealthStatus.HEALTHY
+    warnings: list[HealthWarning] = field(default_factory=list)
+
+    # Performance volatility
+    performance_variance: float = 0.0
+    recent_performance_changes: list[float] = field(default_factory=list)
+
+
+@dataclass
+class HealthMonitorConfig:
+    """Configuration for health monitoring system."""
+
+    enabled: bool = True
+    """Whether health monitoring is enabled."""
+
+    regression_threshold: float = 0.85
+    """Warn if graduated env performance drops below this fraction of graduation level."""
+
+    critical_regression_threshold: float = 0.70
+    """Critical warning if performance drops below this fraction."""
+
+    stagnation_window: int = 100
+    """Number of steps without progress before warning about stagnation."""
+
+    evaluation_frequency: int = 50
+    """How often to evaluate all environments (in training steps)."""
+
+    warning_cooldown: int = 200
+    """Steps before re-issuing the same warning."""
+
+    volatility_threshold: float = 0.15
+    """Coefficient of variation threshold for volatility warnings."""
+
+    performance_history_size: int = 100
+    """Number of recent performance samples to track."""
+
+    prolonged_training_threshold: int = 1000
+    """Warn if environment stays in training for this many steps."""
+
+    enable_wandb_alerts: bool = True
+    """Whether to send alerts to wandb."""
+
+    wandb_alert_rate_limit: int = 300
+    """Minimum seconds between wandb alerts of the same type."""
+
+
+class CurriculumHealthMonitor:
+    """Monitors health of curriculum learning and generates warnings."""
+
+    def __init__(self, config: HealthMonitorConfig):
+        self.config = config
+
+        # Performance tracking
+        self.historical_performance: dict[str, dict[str, deque]] = defaultdict(
```

Comment:
It feels like we're managing many of the same statistics as the curriculum. Should we move some of this logic to the curriculum itself? e.g. when we configure an environment in the CurriculumConfig, we could additionally pass in the monitoring alerts mentioned above?

Maybe sketch that out in a comment/doc or think through the tradeoffs?